pkg/drain/drain_test.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package drain
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	v1meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func getTestSetup(objects []runtime.Object) *fake.Clientset {
+	k8sClient := fake.NewSimpleClientset(objects...)
+	return k8sClient
+}
+
+func dummyDaemonSet(name string) appsv1.DaemonSet {
+	return appsv1.DaemonSet{
+		ObjectMeta: v1meta.ObjectMeta{
+			Name:      name,
+			Namespace: "kube-system",
+		},
+	}
+}
+
+func dummyPod(podMap map[string]string) v1.Pod {
+	pod := v1.Pod{
+		ObjectMeta: v1meta.ObjectMeta{
+			Name:      podMap["name"],
+			Namespace: "kube-system",
+		},
+		Spec: v1.PodSpec{
+			NodeName: "node1",
+		},
+		Status: v1.PodStatus{
+			Phase: v1.PodPhase(podMap["phase"]),
+			ContainerStatuses: []v1.ContainerStatus{
+				{
+					Name:  "container1",
+					Ready: podMap["ready"] == "true",
+				},
+			},
+		},
+	}
+	if podMap["controller"] == "DaemonSet" {
+		pod.SetOwnerReferences([]v1meta.OwnerReference{
+			{
+				Name:       podMap["controllerName"],
+				Kind:       appsv1.SchemeGroupVersion.WithKind("DaemonSet").Kind,
+				Controller: &[]bool{true}[0],
+			},
+		})
+	}
+	return pod
+}
+
+// MakePodList constructs api.PodList from a list of pod attributes
+func makePodList(pods []map[string]string) []runtime.Object {
+	var list []runtime.Object
+	for _, pod := range pods {
+		p := dummyPod(pod)
+		list = append(list, &p)
+	}
+	return list
+}
+
+func TestRollingUpdateDaemonSetMixedPods(t *testing.T) {
```

Comment:
This is a test of the draining code, not of rolling update.

This only tests the ordering of the list returned by `GetPodsForDeletion`. It does not test anything about how that list is handled by `DeleteOrEvictPods`, which is why it failed to catch that the sorting was ineffective.